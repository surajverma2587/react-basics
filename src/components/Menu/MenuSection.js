import React from 'react';
import PropTypes from 'prop-types';
import MenuItem from './MenuItem';

const renderMenuItems = (/* menuItems */) => {
  /* map through the menuItems prop and return a list of MenuItem components */
};

const MenuSection = (/* props object */) => {
  /* bind the data from the props object to the html template */
  /* validate the props and render the component IF valid ELSE null */
  /* remember to call renderMenuItems() to render a list of MenuItem components */
  /* replace the bottom return statement with the populated html template */
  return (
    <div>
      MenuSection Component
    </div>
  );
};

/* display name for the component */
MenuSection.displayName = 'MenuSection';

MenuSection.propTypes = {
  /* add Proptypes here */
};

export default MenuSection;