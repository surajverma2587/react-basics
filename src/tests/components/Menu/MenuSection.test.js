import React from 'react';
import { shallow } from 'enzyme';
import MenuSection from '../../../components/Menu/MenuSection';

const props = {
  id: '11',
  title: 'Section title',
  description: 'Section description',
  menuItems: [
    {
      id: '1',
      title: 'title 1',
      description: 'description 1',
      price: 4
    },
    {
      id: '2',
      title: 'title 2',
      description: 'description 2',
      price: 10
    },
  ],
  onClickMenuItem: jest.fn(),
};

describe('MenuSection Component Tests', () => {
  it('should take snapshot of the component', () => {
    const wrapper = shallow(<MenuSection {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render the component as expected when all props are valid', () => {
    const wrapper = shallow(<MenuSection {...props} />);
    const menuSection = wrapper.find('.menu-section');

    expect(menuSection.exists()).toBe(true);
  });

  it('should render the title of the menu section', () => {
    const wrapper = shallow(<MenuSection {...props} />);
    const title = wrapper.find('h4').at(0);

    expect(title.text()).toEqual(props.title);
  });

  it('should render the description of the menu section', () => {
    const wrapper = shallow(<MenuSection {...props} />);
    const description = wrapper.find('span').at(0);

    expect(description.text()).toEqual(props.description);
  });

  it('should render the collapsible button of the menu section', () => {
    const wrapper = shallow(<MenuSection {...props} />);
    const button = wrapper.find('button');

    expect(button.prop('data-target')).toEqual(`#${props.id}`);
  });

  it('should render the collapsible div of the menu section', () => {
    const wrapper = shallow(<MenuSection {...props} />);
    const collapsibleDiv = wrapper.find('.collapse');

    expect(collapsibleDiv.prop('id')).toEqual(props.id);
  });

  it('should render the MenuItems components', () => {
    const wrapper = shallow(<MenuSection {...props} />);
    const menuItem = wrapper.find('MenuItem');

    expect(menuItem.length).toEqual(2);
  });

  it('should not render the component when props are missing', () => {
    const wrapper = shallow(<MenuSection onClickMenuItem={jest.fn()} />);
    expect(wrapper.find('.menu-section').exists()).toBe(false);
  });

  it('should not render the component when props are invalid', () => {
    const invalidProps = [
      { ...props, id: '' },
      { ...props, title: '' }, 
      { ...props, description: '' },
      { ...props, menuItems: [] },
    ];
    invalidProps.map(invalidProp => {
      const wrapper = shallow(<MenuSection {...invalidProp} />);
      return expect(wrapper.find('.menu-section').exists()).toBe(false);
    });
  });
});
