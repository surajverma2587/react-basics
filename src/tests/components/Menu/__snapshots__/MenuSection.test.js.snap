// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuSection Component when props are valid should render snapshot of the component 1`] = `
<div
  className="menu_section"
>
  <div
    className="menu_section--header"
  >
    <div
      className="row"
    >
      <div
        className="col-sm-11 menu_section--title"
      >
        <span>
          Section title
        </span>
      </div>
      <div
        className="col-sm-1"
      >
        <button
          className="btn btn-primary"
          data-target="#S1"
          data-toggle="collapse"
          type="button"
        >
          v
        </button>
      </div>
    </div>
    <div
      className="row"
    >
      <div
        className="col-sm-12 menu_section--description"
      >
        <span>
          Section description
        </span>
      </div>
    </div>
  </div>
  <div
    className="collapse menu_section--body"
    id="S1"
  >
    <MenuItem
      description="Description 1"
      id="I1"
      key="I1"
      onClick={[MockFunction]}
      price={4.99}
      title="Title 1"
    />
    <MenuItem
      description="Description 2"
      id="I2"
      key="I2"
      onClick={[MockFunction]}
      price={6.49}
      title="Title 2"
    />
  </div>
</div>
`;
